taskcluster/baseclient.py

```diff
@@ -0,0 +1,324 @@
+            # encoded string, which contains a list of tokens separated by '\'.
+            # The first one is the clientId, the second is an int, the third is
+            # url-safe base64 encoded MAC, the fourth is the ext param.
+            # The problem is that the nested url-safe base64 encoded MAC must be
+            # base64 (i.e. not url safe) or server-side will complain.
+
+            # id + '\\' + exp + '\\' + mac + '\\' + options.ext;
+            resource = mohawk.base.Resource(
+                credentials={
+                    'id': clientId,
+                    'key': accessToken,
+                    'algorithm': 'sha256',
+                },
+                method='GET',
+                ext=utils.toStr(self.makeHawkExt()),
+                url=requestUrl,
+                timestamp=expiration,
+                nonce='',
+                # content='',
+                # content_type='',
+            )
+            bewit = mohawk.bewit.get_bewit(resource)
+            return bewit.rstrip('=')
+
+        bewit = genBewit()
+
+        if not bewit:
+            raise exceptions.TaskclusterFailure('Did not receive a bewit')
+
+        u = urllib.parse.urlparse(requestUrl)
+        return urllib.parse.urlunparse((
+            u.scheme,
+            u.netloc,
+            u.path,
+            u.params,
+            u.query + 'bewit=%s' % bewit,
+            u.fragment,
+        ))
+
+    def _hasCredentials(self):
+        """ Return True, if credentials is given """
+        cred = self.options.get('credentials')
+        return (
+            cred and
+            'clientId' in cred and
+            'accessToken' in cred and
+            cred['clientId'] and
+            cred['accessToken']
+        )
+
+    def _makeHttpRequest(self, method, route, payload=None):
+        """ Make an HTTP Request for the API endpoint.  This method wraps
+        the logic about doing failure retry and passes off the actual work
+        of doing an HTTP request to another method."""
+
+        baseUrl = self.options['baseUrl']
+        # urljoin ignores the last param of the baseUrl if the base url doesn't end
+        # in /.  I wonder if it's better to just do something basic like baseUrl +
+        # route instead
+        if not baseUrl.endswith('/'):
+            baseUrl += '/'
+        url = urllib.parse.urljoin(baseUrl, route.lstrip('/'))
+        log.debug('Full URL used is: %s', url)
+
+        hawkExt = self.makeHawkExt()
+
+        # Serialize payload if given
+        if payload is not None:
+            payload = utils.dumpJson(payload)
+
+        # Do a loop of retries
+        retry = -1  # we plus first in the loop, and attempt 1 is retry 0
+        retries = self.options['maxRetries']
+        while retry < retries:
+            retry += 1
+            # if this isn't the first retry then we sleep
+            if retry > 0:
+                snooze = float(retry * retry) / 10.0
```

Comment:
That's verbatim from https://github.com/taskcluster/taskcluster-client.py/blob/master/taskcluster/client.py#L364 .

Once I go through `git blame` and skip past my flake8 whitespace changes, it looks like this is from your commit: https://github.com/taskcluster/taskcluster-client.py/commit/0533cbe8de5623151694e151fbae024049ac2351#diff-cef37b865d30b6dc7000a5e7c28d3fd6R330